README.md
# vredis

A V-Bucket aware redis client for Python.
Allows distribution of Redis data across multiple servers.

## Prerequisites

vredis requires redis-py from Andy McCurdy to be installed first: https://github.com/andymccurdy/redis-py

    $ sudo pip install redis

## Installation

    $ sudo pip install vredis

From source:

    $ sudo python setup.py install

## Getting Started

    >>> import vredis
    >>> r = vredis.VRedis(hosts=[('localhost', 6379, 128), ('localhost', 6379, 255)])
    >>> r.set('foo', 'bar')
    True
    >>> r.get('foo')
    'bar'
    
hosts parameter is a list of tuples, one for each server. Each tuple specifies hostname, port, and ending bucket in the hash ring (out of 255).

The example above has 2 identical hosts using port 6379, first one gets hashes 0..128, second one gets 129..255. 

## Current stage

dbsize, keys, flushall, set, get, mset, and mget have been implemented and tested in VRedis

## TODO

implement and test all the other commands from http://redis.io/commands


CHANGES
* 1.1.0
    * implement dbsize, keys commands
    * better mocking in unit tests
* 1.0
    * initial release, implement get, set, mget, and mset


tests/__init__.py
import unittest
from vredistest import VRedisTest

def all_tests():
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(VRedisTest))
    return suite


tests/mockredis.py
#!/usr/bin/env python
import redis
import re

redis_data = {}

class MockConnectionPool(object):
    def __init__(self, connection_class=redis.Connection, 
                 max_connections=None, **connection_kwargs):
        self.server_str = "%s:%s" % (connection_kwargs['host'], connection_kwargs['port'])
        redis_data[self.server_str] = {}
        self.connection = None
    
    def release(self, connection):
        self.connection = None
        
    def get_connection(self, command_name, *keys, **options):
        if self.connection is None:
            self.connection = MockConnection(self.server_str)
        return self.connection

class MockConnection(object):
    def __init__(self, server_str, host='localhost', port=6379, db=0, password=None,
                 socket_timeout=None, encoding='utf-8',
                 encoding_errors='strict', parser_class=redis.connection.PythonParser):
        self.server_str = server_str
        self.last_result = None
    
    def send_command(self, *args):
        if args[0] == "SET":
            redis_data[self.server_str][args[1]] = args[2]
            self.last_result = "OK"
        elif args[0] == "MSET":
            redis_data[self.server_str].update(zip(args[1::2], args[0::2][1:]))
            self.last_result = "OK"
        elif args[0] == "GET":
            self.last_result = redis_data[self.server_str][args[1]]
        elif args[0] == "MGET":
            self.last_result = [redis_data[self.server_str][x] if x in redis_data[self.server_str] else None for x in args[1:]]
        elif args[0] == "KEYS":
            pattern = args[1].replace("*", ".*") # quick hack
            self.last_result = [x if re.match(pattern, x) else None for x in redis_data[self.server_str]]
        elif args[0] == "DBSIZE":
            self.last_result = len(redis_data[self.server_str])
        elif args[0] == "FLUSHALL":
            redis_data[self.server_str] = {}
            self.last_result = "OK"
    
    def read_response(self):
        return self.last_result


tests/vredistest.py
#!/usr/bin/env python
import unittest
import random
import string
from vredis import VRedis
from mockredis import MockConnectionPool

class VRedisTest(unittest.TestCase):
    def setUp(self):
        # setup VRedis 
        self.vr = VRedis(
            hosts=[
                ('1', 1, 85),
                ('2', 2, 170),
                ('3', 3, 255)
            ],
            cls=MockConnectionPool
        )

    def tearDown(self):
        self.vr.flushall()

    def testhash(self):
        results = {}
        for i in range(256):
            results[i] = 0
        for i in range(256 * 1000):
            # hash a random string
            hash = self.vr.hash(''.join(random.choice(string.ascii_uppercase + string.digits) for x in range(10)))
            results[hash] += 1
        for i in range(256):
            self.assertIn(results[i], range(800, 1200))

    def test_get_server(self):
        s = self.vr.get_server("1") # hash starts with c4 - 196
        self.assertEquals(s, "3:3")
        s = self.vr.get_server("2") # hash starts with c8 - 200
        self.assertEquals(s, "3:3")
        s = self.vr.get_server("3") # hash starts with ec - 236
        self.assertEquals(s, "3:3")
        s = self.vr.get_server("4") # hash starts with a8 - 168
        self.assertEquals(s, "2:2")
        s = self.vr.get_server("6") # hash starts with 16 - 22
        self.assertEquals(s, "1:1")
    
    def test_get(self):
        self.vr.set("test", "1")
        self.assertEquals(self.vr.get("test"), "1")
        self.assertEquals(self.vr.dbsize(), 1)

    def test_mget(self):
        data = {
            "1": 1, # server 3
            "2": 2, # server 3
            "3": 3, # server 3
            "4": 4, # server 2
            "6": 6  # server 1
        }
        self.assertEquals(self.vr.mset(data), True)
        self.assertEquals(self.vr.dbsize(), len(data))
        self.assertEquals(self.vr.mget(data.keys()), data.values())
        for i in data:
            self.assertEquals(self.vr.get(i), data[i])
    
    def test_mget_individual(self):
        data = {
            "1": 1, # server 3
            "2": 2, # server 3
            "3": 3, # server 3
            "4": 4, # server 2
            "6": 6  # server 1
        }
        for i in data:
            self.assertEquals(self.vr.set(i, data[i]), True)
        self.assertEquals(self.vr.mget(data.keys()), data.values())
        for i in data:
            self.assertEquals(self.vr.get(i), data[i])
        self.assertEquals(self.vr.dbsize(), len(data))
    
    def test_keys(self):
        data = {
            "1": 1, # server 3
            "2": 2, # server 3
            "3": 3, # server 3
            "4": 4, # server 2
            "6": 6  # server 1
        }
        self.assertEquals(self.vr.mset(data), True)
        self.assertEquals(self.vr.dbsize(), len(data))
        keys = self.vr.keys("*")
        self.assertEquals(len(keys), len(data.keys()))
        for key in data.keys():
            self.assertTrue(key in keys)
    
if __name__ == '__main__':
    unittest.main()


vredis/vredis.py
#!/usr/bin/env python
import redis
import hashlib
from redis.exceptions import ConnectionError

class VRedis(redis.Redis):
    """ Provides a client-side vbucket-aware interface to Redis """
    
    def __init__(self, hosts, db=0, socket_timeout=None, 
                 charset='utf-8', errors='strict', 
                 cls=redis.ConnectionPool):
        """ Initialize VRedis object """
        super( VRedis, self ).__init__(db=db, socket_timeout=socket_timeout, 
            charset=charset,errors=errors)
        self.ring = []
        self.hosts = hosts
        self.pools = {}
        for host, port, end_bucket in self.hosts: 
            # initialize a connection pool per host
            self.pools['%s:%d' % (host, port)] = cls(host=host, port=port)

        self.generate_ring()

    def generate_ring(self):
        """ Ring size is 255 by default """
        for host in self.hosts:
            (hostname, port, end_bucket) = host
            self.ring.append(('%s:%d' % (hostname, port), end_bucket))

    @staticmethod
    def hash(message):
        """
        Hashing function that returns an int 0..255
        based on the hash of message
        """
        return int('0x%s' % (hashlib.new('md5', message).hexdigest()[:2]), 16)

    def get_server(self, key):
        """ Return the server host:port string to be used for given key """
        position = self.hash(key)
        for host, limit in self.ring:
            if limit >= position:
                return host

    def mget(self, keys, *args):
        """
        Split the keys into lists per server, 
        execute on each server separately, 
        and combine and return the results
        """
        keys = redis.client.list_or_args(keys, args)
        results = {}
        for key in keys:
            hostkey = self.get_server(key)
            if hostkey in results:
                results[hostkey][key] = None
            else:
                results[hostkey] = {key: None}
        # results looks like: 
        # {"server1": 
        #    {key1: None, key2: None, key3: None},
        #  "server2": 
        #    {key4: None, key5: None, key6: None}
        # } 
        for server, server_data in results.iteritems():
            server_keys = server_data.keys()
            new_args = ["MGET"]
            new_args.extend(server_keys)
            values = self.execute_command_on_server(server, *new_args)
            server_data.update(zip(server_keys, values))
        output = []
        for key in keys:
            for server, server_data in results.iteritems():
                if key in server_data:
                    output.append(server_data[key])
                    break
        return output
    
    def dbsize(self):
        """ Run dbsize on all the servers and return the sum """
        result = 0
        for host, limit in self.ring:
            tmp = self.execute_command_on_server(host, 'DBSIZE')
            result = result + tmp
        return result
    
    def keys(self, pattern='*'):
        """ Run keys on all the servers and return the combines set """
        result = []
        for host, limit in self.ring:
            tmp = self.execute_command_on_server(host, 'KEYS', pattern)
            result.extend(tmp)
        return result
    
    def mset(self, mapping):
        """
        Split the keys into dictionaries per server, 
        execute on each server separately, 
        and return True only if all of the servers returned True
        """
        items = []
        for pair in mapping.iteritems():
            items.extend(pair)
        
        keys = {}
        for key, val in mapping.iteritems():
            hostkey = self.get_server(key)
            if hostkey in keys:
                keys[hostkey].extend([key, val])
            else:
                keys[hostkey] = [key, val]
        # keys looks like: 
        # {"server1": 
        #    [key1, val1, key2, val2],
        #  "server2": 
        #    [key3, val3, key4, val4]}
        result = True 
        for server, keys in keys.iteritems():
            tmp = self.execute_command_on_server(server, "MSET", *keys)
            result = result and tmp
        return result

    def execute_command(self, *args, **options):
        """ Picks the correct server and executes command on that server """
        
        if len(args) == 1:
            # execute command on all servers
            # assume we have to return a boolean value
            result = True
            for host, limit in self.ring:
                tmp = self.execute_command_on_server(host, *args, **options)
                result = result and tmp
            return result
        else:
            key = args[1]
            hostkey = self.get_server(key)
            return self.execute_command_on_server(hostkey, *args, **options)
        
    def execute_command_on_server(self, server, *args, **options):
        """
        All of the commands go to a single server now determined by 
        the server parameter
        """
        command_name = args[0]
        pool = self.pools[server]
        connection = pool.get_connection(command_name, **options)
        
        try:
            connection.send_command(*args)
            return self.parse_response(connection, command_name, **options) 
        except ConnectionError:
            connection.disconnect()
            connection.send_command(*args)
            return self.parse_response(connection, command_name, **options)
        finally:
            pool.release(connection)
